design/dpop.md

```diff
@@ -0,0 +1,308 @@
+
+_(OpenID Connect Core 1.0,_ [_5.3. UserInfo Endpoint_][11]_)_
+
+Therefore, UserInfo endpoint must also be made DPoP-aware.
+
+Classes/methods affected:
+
+* org.keycloak.protocol.oidc.endpoints.UserInfoEndpoint
+	* issueUserInfo()
+
+#### Logout Endpoint
+
+The `POST` variant of logout invocation uses refresh token, which could be DPoP-bound. Therefore, Logout endpoint must also be made DPoP-aware.
+
+Classes/methods affected:
+
+* org.keycloak.protocol.oidc.endpoints.LogoutEndpoint
+	* logoutToken()
+
+#### Other Endpoints
+
+The following endpoints in Keycloak are public:
+
+* Server Discovery
+* Server JWK set
+* Authorization
+
+The following endpoints don’t use bearer tokens, but rather client credentials:
+
+* Token Introspection
+* Token Revocation
+
+The following endpoints use cookie authentication:
+
+* Check Session iframe
+
+Therefore, no other endpoints should be affected.
+
+#### Client Registration
+Client registration should support DPoP-related parameters inside client metadata.
+
+Classes/methods affected:
+
+* org.keycloak.services.clientregistration.oidc.DescriptionConverter
+* org.keycloak.representations.oidc.OIDCClientRepresentation
+
+### Models
+
+#### Access Token
+
+New property, `jkt`, should be recognized as a member of the `cnf` claim.
+
+Classes/methods affected:
+
+* org.keycloak.representations.AccessToken.CertConf
+	* rename to Conf or Confirmation; introduce `keyThumbprint` property
+
+#### Client
+
+There are the following DPoP-related client settings:
+
+* DPoP mode (enabled/optional/disabled)
+* DPoP proof lifetime (integer)
+* DPoP clock skew (integer)
+
+Those could be backed by client attributes, hence no changes to the client model and DB schema.
+
+Classes/methods affected:
+
+* org.keycloak.protocol.oidc.OIDCAdvancedConfigWrapper
+
+### Adapters
+
+#### Java Adapters
+
+In the Bearer only mode, Java adapters must:
+
+* enforce DPoP proof validation in the “Enabled/Required” mode;
+* allow for DPoP validation of DPoP-bound tokens in the “Optional” mode;
+* ignore DPoP proofs in the “Disabled” mode.
+
+The following options should be exposed in the adapter configuration:
+
+* DPoP Mode: required/optional/disabled
+* DPoP Proof Lifetime
+* DPoP Allowed Clock Skew
+
+> :question: Should we support DPoP for the “traditional” (monolithic) web applications, where adapters themselves handle Authorization Code flow, and the tokens are not exposed to the user agent?
+
+Classes/methods affected:
+
+*   org.keycloak.adapters.\*
+
+##### Application Clustering
+
+As per the spec,
+
+> Servers SHOULD store, in the context of the request URI, the `jti` value of each DPoP proof for the time window in which the respective DPoP proof JWT would be accepted and decline HTTP requests to the same URI for which the `jti` value has been seen before.
```

Comment:
As you pointed, this is quite challenging to support especially in the cluster environment. I would personally not support this requirement for now (Specification says "SHOULD") unless there is very strong requirement of it. I see there is some security gain, however the price to pay store additional data on the resource servers side is the challenge (especially for the clustering environment).

Reply:
Agreed.